icicle-source/src/Icicle/Source/Type/Subst.hs

```diff
@@ -3,108 +3,54 @@
 --
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE PatternGuards #-}
+{-# LANGUAGE PatternGuards     #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TupleSections     #-}
 module Icicle.Source.Type.Subst (
     SubstT
   , substT
   , substC
-  , substFT
   , compose
   , unifyT
   ) where
 
 
 import                  Icicle.Common.Base
+
 import                  Icicle.Source.Type.Base
 import                  Icicle.Source.Type.Compounds
 
 import                  P
 
+import                  Control.Lens (over)
+
 import qualified        Data.Map as Map
 import qualified        Data.Set as Set
 import                  Data.Hashable (Hashable)
 
 type SubstT n = Map.Map (Name n) (Type n)
 
+
+--- | Substitute into a type.
 substT :: Eq n => SubstT n -> Type n -> Type n
-substT ss tt
- = canonT
- $ go tt
+substT ss
+ = canonT . go ss
  where
-
-  go t
-   = case t of
-      BoolT         -> t
-      TimeT         -> t
-      DoubleT       -> t
-      IntT          -> t
-      StringT       -> t
-      UnitT         -> t
-      ErrorT        -> t
-
-      ArrayT  a     -> ArrayT  (go a)
-      GroupT  a b   -> GroupT  (go a) (go b)
-      OptionT a     -> OptionT (go a)
-      PairT   a b   -> PairT   (go a) (go b)
-      SumT    a b   -> SumT    (go a) (go b)
-      StructT fs    -> StructT (Map.map go fs)
-
-      Temporality a b       -> Temporality (go a) (go b)
-      TemporalityPure       -> t
-      TemporalityElement    -> t
-      TemporalityAggregate  -> t
-
-      Possibility a b       -> Possibility (go a) (go b)
-      PossibilityPossibly   -> t
-      PossibilityDefinitely -> t
-
-      TypeVar n
-       | Just t' <- Map.lookup n ss
-       -> t'
-       | otherwise
-       -> t
-
+  go ss0 = \case
+    TypeVar n
+      | Just t' <- Map.lookup n ss
+      -> t'
+    TypeForall ns cs r
+      -> let ss1  = ss Map.\\ Map.fromList ((,()) <$> ns)
+             cs1  = fmap (substC ss1) cs
+             r1   = go ss1 r
```

Comment:
Bit tricky here. I think you need to check whether any of `ns` occur in the free variables of the payloads of `ss` (something like `unions (map freeT $ Map.values ss)`). If so, you need to rename `ns`. (Which is a bit annoying because this function doesn't have a name supply at the moment)

Example is something like
```
substT (a := (b -> b)) (forall b. a -> b)
-->
forall c. (b -> b) -> c
```
Here, the `b` in the substitution payload is a free variable, but the `b` in the target type is bound by the forall

Reply:
It might be the case that because of the form of foralls and how substitutions are generated, this will never occur in practice.. I'm not sure, have to think about that - if so, a big warning might suffice

Reply:
I would imagine that's the case, as the only way we generate these is when building the functions with a fresh name supply, and they are strictly non-recursive.

I definitely see your point though.